### DISCUSSION TITLE
How to upload/download a directory quickly using python SDK?

### DISCUSSION THREAD

**Discussion by sjtu-biggie** [minio/minio#11432] [Q&A] [answered]
When using mc commands to upload a directory with many small files:
`./mc cp --recursive test/small_file_1024 local/test`
It took within a second to finish.

However, if I use `fput_object()` or `put_object()`, which is provided by python-SDK, I have to iterate the directory and upload the file one by one, taking ten times as long as the previous method.
Is there any way to optimize the code? If not, can I just use `mc` in the `python` script without caring about thread-safe problem?

**Comment by harshavardhana** [comment, marked as answer]
> However, if I use `fput_object()` or `put_object()`, which is provided by python-SDK, I have to iterate the directory and upload the file one by one, taking ten times as long as the previous method.
> Is there any way to optimize the code? If not, can I just use `mc` in the `python` script without caring about thread-safe problem?

minio-py is SDK, not a tool - `mc` uses `minio-go` library to implement `cp` functionality. If you wish to write the same thing with Python you should write the following pseudo code should be the way to do it

```
entries = os.readdir(...)
for entry in range entries:
      fput_object()
```

**Reply by sjtu-biggie** [reply]
Here is my code,uploading a directory with 1024 files with total size 128MB.
```python
local = MinioClient(MINIO_URL, ACCESS_KEY, SECRET_KEY)
start = time.time()
os.system("./mc cp --recursive test/small_file_1024 local/test")
end = time.time()
print("times 1 {}".format(end - start))

os.system("./mc rm --recursive --force local/test/small_file_1024 &> /dev/null")

start = time.time()
dir = 'test/small_file_1024'
for file in os.listdir(dir):
    local.client.fput_object('test', 'small_file_1024/' + file, 'test/small_file_1024/' + file)
end = time.time()
print("times 2 {}".format(end - start))
```
And the result shows:
```
times 1 0.35630035400390625
times 2 3.7435789108276367
```

**Reply by harshavardhana** [reply]
This is because `mc` has concurrency built-in @sjtu-biggie - you need to use python threading library to concurrently upload content.

**Reply by sjtu-biggie** [reply]
It helps. Much thanks!

**Reply by sjtu-biggie** [reply]
Sorry to disturb you again, I used `threading` and `threadpool` library to improve the concurrency, now it took only 3-4 times as long as `mc cp --recursive`. But it's obviously not the best practice.I'm wondering how to reach the best performance in this scenario (using python)?
Here's my implementation.
```python
pool = threadpool.ThreadPool(8)
start = time.time()
dir = 'test/small_file_1024'
requests = threadpool.makeRequests(upload, os.listdir(dir))
[pool.putRequest(req) for req in requests]
pool.wait()
end = time.time()
```

**Reply by harshavardhana** [reply]
@balamurugana - would you be able to help?
